File: WebApps/AngularClient/Services/CatalogService.cs
using ApiClient.Catalog.ApiClient;
using ApiClient.Catalog.Models;

namespace AngularClient.Services;

public interface ICatalogService
{
    Task<List<ProductSummary>?> GetProductsAsync(CancellationToken cancellationToken = default);
    Task<ProductDetail> GetProductById(string id, CancellationToken cancellationToken = default);
    Task<List<ProductSummary>?> GetProductByCategoryAsync(string category, CancellationToken cancellationToken = default);
    Task<ProductDetail?> CreateProductAsync(CreateProductRequestBody requestBody, CancellationToken cancellationToken = default);
    Task<ProductDetail?> DeleteProductAsync(string id, CancellationToken cancellationToken = default);
    Task<ProductDetail?> UpdateProductAsync(UpdateProductRequestBody requestBody, CancellationToken cancellationToken);
}

public class CatalogService : ICatalogService
{
    private readonly ICatalogApiClient _catalogApiClient;

    public CatalogService(ICatalogApiClient catalogApiClient)
    {
        _catalogApiClient = catalogApiClient;
    }

    public async Task<List<ProductSummary>?> GetProductsAsync(CancellationToken cancellationToken)
    {
        var result = await _catalogApiClient.GetProducts(cancellationToken);

        if (result.IsSuccessCode && result.Data is not null)
        {
            return result.Data;
        }

        return null;
        //Ocelot ---> CatalogAPI
    }

    public async Task<ProductDetail> GetProductById(string id, CancellationToken cancellationToken)
    {
        var result = await _catalogApiClient.GetProductByIdAsync(id, cancellationToken);
        if (result.IsSuccessCode && result.Data is not null)
        {
            return result.Data;
        }

        return null;
    }

    public async Task<List<ProductSummary>?> GetProductByCategoryAsync(string category, CancellationToken cancellationToken)
    {
        var result = await _catalogApiClient.GetProductByCategoryAsync(category, cancellationToken);
        
        if (result.IsSuccessCode && result.Data is not null)
        {
            return result.Data;
        }

        return null;
    }

    public async Task<ProductDetail> CreateProductAsync(CreateProductRequestBody requestBody, CancellationToken cancellationToken)
    {
        var result = await _catalogApiClient.CreateProductAsync(requestBody, cancellationToken);
        if (result.IsSuccessCode && result.Data is not null)
        {
            return result.Data;
        }

        return null;
    }

    public async Task<ProductDetail> DeleteProductAsync(string id, CancellationToken cancellationToken)
    {
        var result = await _catalogApiClient.DeleteProductAsync(id, cancellationToken);  
        if(result.IsSuccessCode && result.Data is not null)
        { 
            return result.Data; 
        }

        return null;
    }

    public async Task<ProductDetail> UpdateProductAsync (UpdateProductRequestBody requestBody, CancellationToken cancellationToken)
    {
        var result = await _catalogApiClient.UpdateProductAsync(requestBody, cancellationToken);
        if(result.IsSuccessCode && result.Data is not null)
        {
            return result.Data;
        }
        return null;
    }
}

File: WebApps/AngularClient/ClientApp/src/app/core/service/catalog.service.ts
import {HttpParams} from "@angular/common/http";
import {Injectable} from "@angular/core";
import {environment} from "src/environments/environment";
import {ApiService} from "./api.service";
import { ProductDetail } from "../models/product-detail.model";
import { ProductSummary } from "../models/product-summary.model";

@Injectable()
export class CatalogService {
  apiName: string = "Catalog";

  constructor(public apiService: ApiService) {

  }

  async getProductsAsync(): Promise<ProductSummary[]> {
    return await this.apiService.getAsync(`${environment.baseApiUrl}/${this.apiName}/GetProducts`, new HttpParams());
  }

  async getProductByIdAsync(id: string): Promise<ProductDetail> {
    return await this.apiService.getAsync(`${environment.baseApiUrl}/${this.apiName}/GetProductById/${id}`, new HttpParams());
  }

  async getProductByCategoryAsync(category: string): Promise<ProductSummary> {
    return this.apiService.getAsync(`${environment.baseApiUrl}/${this.apiName}/GetProductByCategory/${category}`, new HttpParams());
  }

  async createProductAsync(catalogDetail: any): Promise<ProductDetail> {
    return this.apiService.postAsync(`${environment.baseApiUrl}/${this.apiName}/CreateProduct`, catalogDetail, new HttpParams());
  }

  async updateProductAsync(catalogDetail: any): Promise<ProductDetail> {
    return this.apiService.postAsync(`${environment.baseApiUrl}/${this.apiName}/UpdateProduct`, catalogDetail, new HttpParams());
  }

  async deleteProductAsync(id: string): Promise<boolean> {
    return await this.apiService.postAsync(`${environment.baseApiUrl}/${this.apiName}/DeleteProduct`, id, new HttpParams());
  }
}
